src/sort_transform.c

```diff
@@ -289,20 +290,55 @@ sort_transform_ec(PlannerInfo *root, EquivalenceClass *orig)
 				newec->ec_derives = list_copy(orig->ec_derives);
 				newec->ec_relids = bms_copy(orig->ec_relids);
 				newec->ec_has_const = orig->ec_has_const;
-				newec->ec_has_volatile = orig->ec_has_volatile;
+
+				/* even if the original EC has volatile (it has time_bucket_gapfill)
+				 * this ordering is purely on the time column, so it is non-volatile
+				 * and should be propagated to the children
+				 */
+				newec->ec_has_volatile = false;
 				newec->ec_below_outer_join = orig->ec_below_outer_join;
 				newec->ec_broken = orig->ec_broken;
 				newec->ec_sortref = orig->ec_sortref;
 				newec->ec_merged = orig->ec_merged;
-			}
 
+				/* volatile ECs only ever have one member, that of the root,
+				 * so if the original EC was volatile, we need to propagate the
+				 * new EC to the children ourselves
+				 */
+				propagate_to_children = orig->ec_has_volatile;
+				/* even though time_bucket_gapfill is marked as VOLATILE to
+				 * prevent the planner from removing the call, it's still safe
+				 * to use values from child tables in lieu of the output of the
+				 * root table. Among other things, this allows us to use the
+				 * sort-order from the child tables for the output
+				 */
+				orig->ec_has_volatile = false;
```

Comment:
Am I correct that the added test will catch the case when PG query planner will remove the call? It's good to have such test if next version of PG will change the planning logic.

Reply:
yeah, pretty much all our gapfill tests will break without it